### PR TITLE
Support dynamic shape

### PR DESCRIPTION
**Pull request by dayo09**
Let's add dynamic shape example and test using onert.

TICO-DCO-1.0-Signed-off-by: Dayoung Lee <dayoung.lee@samsung.com>

### CODE REVIEW

File: test/modules/net/mlp_dyn.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        )
+        return down_proj
+
+    def get_example_inputs(self):
+        # (B, Seq_Len, Dim)
+        return (torch.randn(B, SEQ_LEN, DIM),)
+
+    def get_dynamic_shapes(self):
+        batch = Dim("batch", min = 1, max = 128)
+        dynamic_shapes = {"x": {0: batch},}
```
Comment (dayo09) [inline test/modules/net/mlp_dyn.py:43]: ### Exported Circle Model ✔️ 

image

File: test/pt2_to_circle_test/builder.py [hunk 2]
```diff
@@ -150,7 +150,7 @@ def _run(
         if without_inference:
             return
 
-        USE_ONERT = os.environ.get("CCEX_RUNTIME") == "onert" or dynamic
+        USE_ONERT = os.environ.get("CCEX_RUNTIME") == "onert"
```
Comment (dayo09) [inline test/pt2_to_circle_test/builder.py:153]: Let's set `use_onert` explicitly

File: tico/utils/dtype.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    np.dtype("int64"): torch.int64,
+    np.dtype("int32"): torch.int32,
+    np.dtype("int16"): torch.int16,
+    np.dtype("int8"): torch.int8,
+    np.dtype("uint8"): torch.uint8,
+    np.dtype("bool"): torch.bool,
+}
+
+
+def numpy_dtype_to_torch_dtype(np_dtype):
```
Comment (jinevening) [inline tico/utils/dtype.py:19]: Could you add type hints?

```suggestion
def numpy_dtype_to_torch_dtype(np_dtype):
```

File: tico/utils/dtype.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    return NUMPY_TO_TORCH_DTYPE_DICT[np.dtype(np_dtype)]
```
Comment (jinevening) [inline tico/utils/dtype.py:20]: Isn't it possible to just pass np_dtype?

```suggestion
    return NUMPY_TO_TORCH_DTYPE_DICT[np_dtype]
```

File: test/modules/op/depthwise_conv2d.py [hunk 5]
```diff
@@ -29,6 +32,28 @@ def get_example_inputs(self):
         return (torch.randn(1, 8, 64, 64),)
 
 
+@tag.use_onert
+class SimpleDepthwiseConvDynamicShape(torch.nn.Module):
```
Comment (dayo09) [inline test/modules/op/depthwise_conv2d.py:36]: @jinevening Thanks for the comment! I added the logic to add_tensor_from_scratch.

image

This test ensures that any intermediate tensor made by `add_tensor_from_scratch` has a dynamic shape - padding in this case.

File: test/pt2_to_circle_test/builder.py [hunk 6]
```diff
@@ -187,15 +187,6 @@ def build(self, submodule):
...
-            testdict[base_name] = NNModuleTest(base_name, module_instance).make(
-                dynamic=False
-            )
-
-            if hasattr(module_instance, "get_dynamic_shapes"):
-                dyn_name = base_name + "_dynamic"
-                testdict[dyn_name] = NNModuleTest(dyn_name, module_instance).make(
-                    dynamic=True
-                )
+            testdict[base_name] = NNModuleTest(base_name, module_instance).make()
```
[the remark this answers is not in the document]
Comment (mhs4670go) [reply test/pt2_to_circle_test/builder.py:190]: @dayo09 The reason why I branched was because I wanted to reuse the existing tests. For example, if I introduce a `get_dynamic_shapes` to existing tests, the test would be run twice: one with `get_example_inputs` and the other with `get_dynamic_shapes`. 

But, with current approach, `SimpleDepthwiseConvDynamicShape` seems to be run only once — with `get_dynamic_shapes`. Is it what you intended?

File: tico/serialize/circle_graph.py [hunk 7]
```diff
@@ -241,6 +250,15 @@ def add_tensor_from_scratch(
         if source_node is not None:
             self.name_to_node[tensor.name] = source_node
         tensor.shape = shape
+
+        # Handle dynamic shape
+        if any(isinstance(s, torch.SymInt) for s in tensor.shape):
```
Comment (jinevening) [inline tico/serialize/circle_graph.py:255]: The type hints and comments of this function have to be updated as well.

```python
        shape: List[int],
```

```
        shape : List[int]
            The shape of the tensor.
```

**Review by jinevening — APPROVED** [review]
LGTM w/ minor comment

File: tico/serialize/circle_graph.py (hunk 7 above)
Comment (dayo09) [reply tico/serialize/circle_graph.py:255]: Let me revisit there in the next PR!
